stan/math/prim/fun/identity_constrain.hpp

```diff
@@ -30,12 +30,13 @@ inline T identity_constrain(const T& x) {
  * <p>This method is effectively a no-op and mainly useful as a
  * placeholder in auto-generated code.
  *
- * @tparam T type of scalar.
+ * @tparam T type of scalar
+ * @tparam S type of log probability
  * @param[in] x scalar
  * @return transformed input
  */
-template <typename T>
-inline T identity_constrain(const T& x, T& /*lp*/) {
+template <typename T, typename S>
+inline T identity_constrain(const T& x, S& lp) {
```

Comment:
This should still be `/*lp*/` as the `lp` variable is unused in the function body and it may produce a warning.

Reply:
We already have a version without the `lp` argument. This other one is for symmetry with the other constraint functions which have that argument and pass it to `identity_constraint`, even if it's not used. I don't feel I understand this stuff well enough to depart from the existing signatures.

Reply:
Ok. I don't think symmetry alone is a good enough reason to keep a piece of code. @bob-carpenter, I think you understand this better than I do -- is there any good reason for overloading `identity_constraint` here?

Reply:
> This should still be /*lp*/ as the lp variable is unused in the function body

If the types get relaxed, we need it for type inference.  That's done statically, so there's no runtime overhead.

> I don't feel I understand this stuff well enough to depart from the existing signatures.

I built the constrain/unconstrain idiom for doing this when we first built the language.  That was very early on in my understanding of C++ and we were having a lot of trouble getting the simplest things to compile.

The current change strictly generalizes the set of instantiations allowed.  In Stan as currently deployed, all arguments will be at least type `double`.  

> is there any good reason for overloading identity_constraint here?

The goal is to relax the type instantiations here to allow for `int` arguments to constraints with `double` or `var` or `fvar` arguments to `lp`.  That will come up if these ever get exposed to the language, as we've been proposing to do forever. 

As is, we can't write tests with `int` arguments along with `double` arguments for `lp`.  

I opened issue https://github.com/stan-dev/math/issues/1733 to request a universal fix.

Whoever's reviewing this needs to decide if it's OK to have a partial solution.  I've been of a fix-it-as-you-go mindset with this kind of stuff, but I can see the argument for consistency across constraints and it's up to whoever is reviewing to decide.

Thanks for looking at this so closely, by the way!

Reply:
I'm ok with the the fix-it-as-you-go approach, especially since this PR is straightforward and we have an issue for the universal fix.